### PR TITLE
Make label and field selector query strings versionable.

### PR DESCRIPTION
**Pull request by brendandburns**
@bgrant0607 

See https://github.com/GoogleCloudPlatform/kubernetes/issues/2131 for some context.

I know the unit tests are broken.  If this LGTM to folks, I will fix them.

### CODE REVIEW

**Comment by bgrant0607** [comment]
cc @lavalamp

**Comment by bgrant0607** [comment]
Thanks. API change LGTM. I'm in favor of renaming the query parameters to "label-selector" and "field-selector". I suppose reverse lookup would be "selector-target-labels" or somesuch.

Delegating detailed review to others.

We should also clarify kubectl flags at some point.

**Comment by bgrant0607** [comment]
BTW, bonus points for making these consistent across list operations and/or documented using go-restful.

File: pkg/client/fake.go [hunk 1]
```diff
@@ -124,19 +124,19 @@ type FakeRESTClient struct {
 }
 
 func (c *FakeRESTClient) Get() *Request {
-	return NewRequest(c, "GET", &url.URL{Host: "localhost"}, c.Codec, c.Legacy, c.Legacy)
+	return NewRequest(c, "GET", &url.URL{Host: "localhost"}, "v1beta1", c.Codec, c.Legacy, c.Legacy)
```
Comment (nikhiljindal) [inline pkg/client/fake.go]: Will recommend using testapi.Version() or latest.Version rather than hardcoding v1beta1

File: pkg/client/request_test.go [hunk 2]
```diff
@@ -242,7 +234,7 @@ func TestTransformResponse(t *testing.T) {
 		{Response: &http.Response{StatusCode: 200, Body: ioutil.NopCloser(bytes.NewReader(invalid))}, Data: invalid},
 	}
 	for i, test := range testCases {
-		r := NewRequest(nil, "", uri, testapi.Codec(), true, true)
+		r := NewRequest(nil, "", uri, "v1beta1", testapi.Codec(), true, true)
```
Comment (nikhiljindal) [inline pkg/client/request_test.go]: Use testapi.Version() instead of hardcoding v1beta1

File: pkg/apiserver/resthandler.go [hunk 3]
```diff
@@ -82,7 +82,13 @@ func GetResource(r RESTGetter, ctxFn ContextFunc, namer ScopeNamer, codec runtim
 }
 
 func parseSelectorQueryParams(query url.Values, version, apiResource string) (label, field labels.Selector, err error) {
-	label, err = labels.Parse(query.Get("labels"))
+	var labelString string
+	if version == "v1beta1" || version == "v1beta2" {
+		labelString = query.Get("labels")
+	} else {
+		labelString = query.Get("label-selector")
```
Comment (nikhiljindal) [inline pkg/apiserver/resthandler.go]: This logic is right now spread at 2 places: here and in pkg/client/request.go.
How about adding a function that takes the apiVersion and returns the appropriate QueryParamName?
We can then use the same function while documenting the query param with go-restful.

Comment (bgrant0607) [inline pkg/apiserver/resthandler.go]: This kind of version test is not acceptable. Among other problems, it would break Openshift. It needs to be abstracted away. I don't think that parameterizing a function by version is the right thing, either.

@smarterclayton Which interface needs to be extended?

Comment (bgrant0607) [inline pkg/apiserver/resthandler.go]: Otherwise, we could add this change to the list of breaking changes we'd like to make.

The client parameterization looks ugly.

Comment (brendandburns) [inline pkg/apiserver/resthandler.go]: This code isn't intended to have a long lifespan.  Instead it is intended to patch the existing behavior until we can turn down v1beta1/v1beta2.

Building in a generic architecture for changing these flags seems overkill for this PR.  (I'm happy to centralize the function in a single place, though)

Comment (lavalamp) [inline pkg/apiserver/resthandler.go]: +1 for centralization; consider a throwing a simple map somewhere that takes version & your FIELD_SELECTOR constants.

Comment (bgrant0607) [inline pkg/apiserver/resthandler.go]: Tests don't use the same version strings, either. This approach is just broken. 

The apiserver is supposed to be generic. The check can't be in pkg/apiserver. 

If you want to just pass strings for the parameters from pkg/master, that could work.

File: pkg/client/request_test.go (hunk 2 above)
Comment (bgrant0607) [inline pkg/client/request_test.go]: Examples to follow:
https://github.com/GoogleCloudPlatform/kubernetes/blob/3f28badae31b935aae72d33d63220923e2cac162/pkg/controller/replication_controller_test.go#L39

File: pkg/client/request.go [hunk 4]
```diff
@@ -287,7 +265,16 @@ func (r *Request) SelectorParam(paramName string, s labels.Selector) *Request {
 	if s.Empty() {
 		return r
 	}
-	return r.setParam(paramName, s.String())
+	name := paramName
+	if r.apiVersion == "v1beta1" || r.apiVersion == "v1beta2" {
```
Comment (bgrant0607) [inline pkg/client/request.go]: https://github.com/GoogleCloudPlatform/kubernetes/blob/3f28badae31b935aae72d33d63220923e2cac162/pkg/client/client.go#L149

File: pkg/client/fake.go (hunk 1 above)
Comment (brendandburns) [inline pkg/client/fake.go]: done.

File: pkg/client/request_test.go (hunk 2 above)
Comment (brendandburns) [inline pkg/client/request_test.go]: done.

File: pkg/apiserver/resthandler.go (hunk 3 above)
Comment (brendandburns) [inline pkg/apiserver/resthandler.go]: ok, moved this to a function in ```api``` and refactored all accesses in terms of a versioned request to that function.

File: pkg/client/request.go (hunk 4 above)
Comment (brendandburns) [inline pkg/client/request.go]: moved to ```api``` and used.

**Comment by brendandburns** [comment]
Comments addressed.  Please take a look.

File: pkg/client/request.go [hunk 5]
```diff
@@ -110,6 +110,8 @@ type Request struct {
 	selector     labels.Selector
 	timeout      time.Duration
 
+	apiVersion string
```
Comment (nikhiljindal) [inline pkg/client/request.go]: This does not seem to be used anywhere.
Is this change required? Or am I missing something.

**Comment by nikhiljindal** [comment]
LGTM.
@bgrant0607 Is this fine now (now that we are using api.LabelSelectorQueryParam()) or do you still think  that we should be passing the strings to apiserver from pkg/master.

**Comment by bgrant0607** [comment]
LGTM for now.

**Comment by smarterclayton** [comment]
I'm currently fixing and refactoring some of this, but I don't think we should be using "field-selector".  In brief, we already use "resourceVersion", I'd like this to be consistent with our field naming ("fieldSelector" would be the JSON name), we're using camelCase in field labels.

If there's no disagreement, I'm going to change this in #5763 where I'm standardizing query parsing to follow our JSON pattern (so you can call api.Scheme.Convert(URL.Query.Values(), &api.Object) and version them).  That will reduce our number of conversion patterns down to 1, and then we don't need special field naming for these sorts of cases (instead, we are deserializing query parameters into a versional object, then converting it to the internal form).

**Comment by derekwaynecarr** [comment]
@smarterclayton +1

**Comment by bgrant0607** [comment]
SGTM. Presumably `s/label-selector/labelSelector/`, also
